skimage/filters/_multiotsu.pyx

```diff
@@ -6,101 +6,295 @@ import numpy as np
+                                            thresh_count=thresh_count,
+                                            sigma_max=sigma_max,
+                                            current_indices=current_indices,
+                                            thresh_indices=thresh_indices)
+
+    else:
+
+        sigma = (_get_var_btwclas(P, S, 0, current_indices[0])
+                 + _get_var_btwclas(P, S,
+                                    current_indices[thresh_count-1]+1,
+                                    nbins-1))
+        for idx in range(thresh_count-1):
+            sigma += _get_var_btwclas(P, S, current_indices[idx]+1,
+                                      current_indices[idx+1])
+        if sigma > sigma_max:
+            sigma_max = sigma
+            thresh_indices[:] = current_indices[:]
 
-    return py_aux_thresh
+    return sigma_max
 
 
-cdef double _find_best_rec(double[:, ::1] var_btwcls, cnp.intp_t min_val,
-                           cnp.intp_t max_val, Py_ssize_t[::1] idx_tuple,
-                           cnp.intp_t divisions, cnp.intp_t depth,
-                           double max_sigma, Py_ssize_t[::1] aux_thresh) nogil:
+cdef void _set_var_btwcls_lut(float [::1] prob,
+                              Py_ssize_t nbins,
+                              float [::1] H,
+                              float [::1] P,
+                              float [::1] S) nogil:
+    """Build the between classes variance lookup table.
+
+    The between classes variance are stored in H. P and S are buffers
+    for storing the first row of respectively the zeroth and first
+    order moments lookup table (see [1]).
+
+    Parameters
+    ----------
+    prob: 2D array
+        Intensities probabilies.
+    nbins: int
+        The number of intensity values.
+    H: (nbins*(nbins + 1) /2) 1D array
+        The Upper triangular part of the between classes variance
+        lookup table.
+    P: (nbins, ) 1D array
+        First row of the zeroth order moments LUT (see [1]).
+    S: (nbins, ) 1D array
+        First row of the first order moments LUT (see [1]).
+
+    References
+    ----------
+    .. [1] Liao, P-S., Chen, T-S. and Chung, P-C., "A fast algorithm for
+           multilevel thresholding", Journal of Information Science and
+           Engineering 17 (5): 713-727, 2001. Available at:
+           <http://ftp.iis.sinica.edu.tw/JISE/2001/200109_01.pdf>
+
     """
-    Recursive function for calculating max_sigma.
+    cdef cnp.intp_t i, j, idx
+    cdef float Pij, Sij
+
+    idx = 1
+    P[0] = prob[0]
+    S[0] = prob[0]
+    for i in range(1, nbins):
+        P[i] = P[i-1] + prob[i]
+        S[i] = S[i-1] + i*prob[i]
+        if P[i] > 0:
+            H[idx] = (S[i]**2)/P[i]
+        idx += 1
+
+    for i in range(1, nbins):
+        for j in range(i, nbins):
+            Pij = P[j] - P[i-1]
+            if Pij > 0:
+                Sij = S[j] - S[i-1]
+                H[idx] = (Sij**2)/Pij
+            idx += 1
```

Comment:
is `idx` restarted or used the last value from the previous cycle?

Reply:
It uses the last value from the previous cycle since it is the current index in `H`

Reply:
but the cycle has fix number f iteration (there is no conditional break)

Reply:
Nop, we loop over the upper triangular part of the matrix, so `idx = nbins * (nbins + 1) / 2`. The difficulty is that `H` here stores only the upper triangular part  of the look up table.

Reply:
so you know the value of `idx`

Reply:
Yes I can compute it for any `i` or `j`, it is what the `_get_var_btwclas_lut` does.

Reply:
I don't know if it is clear, but `H` here is a 1D view of the upper triangular part of the LUT containing the different variance between classes. Since I only store the upper triangular part, I have to match between the 2D coordinates `i, j` an the 1D coordinate `idx`.
When I am setting `H`, I don't really care about the 2D indices (that's why I am simply incrementing `idx`), it is only when I want to access them that I have to "translate" `i, j` into `idx`. That's what `_get_var_btwclas_lut` does.

Reply:
@rfezzani that's definitely not immediately clear and probably something that should be written in as comments in the code. =)

Reply:
You are right, I only exposed this in my comments [1](https://github.com/scikit-image/scikit-image/issues/3975#issuecomment-534889441) and [2](https://github.com/scikit-image/scikit-image/issues/3975#issuecomment-537046656) in the issue #3975, sorry.
I also notice that multiple Cython functions have no docstrings. I will fix that.

Reply:
I added some doc to Cython function ;-)